main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
+        Iterator(rc, iterF)
 
     ///
     /// Returns an empty iterator.
     ///
-    pub def new(_: Region[r]): Iterator[a, r] \ { Read(r), Write(r) } =
-        let done = () -> unsafe_cast true as _ \ Read(r);
-        let next = () -> unsafe_cast bug!("Empty Iterator!") as _ \ Read(r);
-        unsafe_cast Iterator(done, next) as _ \ { Read(r), Write(r) }
+    pub def empty(rc: Region[r]): Iterator[a, ef, r] =
+        let f = () -> unsafe_cast Done as _ \ ef;
+        Iterator(rc, f)
 
     ///
     /// Returns an iterator containing only a single element, `x`.
     ///
-    pub def singleton(r: Region[r], x: a): Iterator[a, r] \ Write(r) =
-        let d = ref false @ r;
-        let done = () -> deref d;
-        let next = () -> {
-            if (done()) {
-                bug!("Empty Iterator!")
-            }
-            else {
-                d := true;
-                x
-            }
-        };
-        Iterator(done, next)
+    pub def singleton(rc: Region[r], x: a): Iterator[a, ef, r] \ Write(r) =
+        repeat(rc, 1, x)
 
     ///
     /// Returns `Some(x)` if `iter` is not empty. Returns `None` otherwise.
     ///
     /// Consumes the head element of `iter`.
     ///
-    pub def next(iter: Iterator[a, r]): Option[a] \ Read(r) =
-        let Iterator(done, next) = iter;
-        if (done())
-            None
-        else
-            Some(next())
-
-    ///
-    /// Returns true if the iterator is empty.
-    ///
-    /// Does *not* consume any elements of the iterator.
-    ///
-    pub def isEmpty(iter: Iterator[a, r]): Bool \ Read(r) =
-        let Iterator(done, _) = iter;
-        done()
+    pub def next(iter: Iterator[a, ef, r]): Option[a] \ { ef, Read(r) } =
+        let Iterator(_, iterF) = iter;
+        def loop() = match iterF() {
+            case Ans(a) => Some(a)
+            case Skip   => loop()
+            case Done   => None
+        };
+        loop()
 
     ///
     /// Returns an iterator of all integers between `b` (inclusive) and `e` (exclusive).
     ///
     /// Returns an empty iterator if `b >= e`.
     ///
-    @Lazy
-    pub def range(r: Region[r], b: Int32, e: Int32): Iterator[Int32, r] \ Write(r) =
-        let i = ref b @ r;
-        let done = () -> deref i >= e;
-        let next = () -> {
-                let res = deref i;
-                if (res < e) {
-                    i := res + 1;
-                    res
-                }
-                else {
-                    bug!("Empty Iterator!")
-                }
+    pub def range(rc: Region[r], b: Int32, e: Int32): Iterator[Int32, ef, r] \ Write(r) =
+        let pos = ref b @ rc;
+        if (e <= b)
+            empty(rc)
+        else {
+            let f = () -> match (deref pos) {
+                case i if i < e => pos := i+1; Ans(i)
+                case _          => Done
             };
-        Iterator(done, next)
+            let iterF = () -> unsafe_cast f() as _ \ { ef, r };
+            Iterator(rc, iterF)
```

Comment:
Here the problem is seemingly different-- because you want `ef` in the signature, but of course there is no user-defined effect (yet!).

Reply:
Thus the cast is required here, but once we have a proper safe cast thing that can be used instead.

Reply:
I think this was either `range` or `repeat` that now return an iterator with "region access" as the effect rather than a polymorphic effect. With "region access" for the effect they don't need casting.